sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetPartitionReaderFactory.scala

```diff
@@ -88,12 +88,18 @@ case class ParquetPartitionReaderFactory(
   private def getFooter(file: PartitionedFile): ParquetMetadata = {
     val conf = broadcastedConf.value.value
     val filePath = file.toPath
+    val split = new FileSplit(file.toPath, file.start, file.length, Array.empty[String])
 
-    if (aggregation.isEmpty) {
-      ParquetFooterReader.readFooter(conf, filePath, SKIP_ROW_GROUPS)
-    } else {
+    if (aggregation.isDefined) {
       // For aggregate push down, we will get max/min/count from footer statistics.
       ParquetFooterReader.readFooter(conf, filePath, NO_FILTER)
+    } else if (enableVectorizedReader) {
+      // This can avoid reading the footer twice(currently only optimize for vectorized read).
+      // Read all the row groups in advance and filter the row groups later if there are
+      // filters that need push down.
+      ParquetFooterReader.readFooter(conf, split, ParquetFooterReader.WITH_ROW_GROUPS)
+    } else {
```

Comment:
it looks we can, since when aggregation push down is enabled, `ParquetScan.isSplitable` returns false, and we'll always read all the row groups in the file, so `NO_FILTER` is the same as `WITH_ROW_GROUPS`.